Gutenboarding: Add separate entrypoint

ockham opened the pull request:
#### Changes proposed in this Pull Request

Branched from @jsnajdr's #37071 (thanks Jarda!). ~Requires #37137 to be merged first.~ ✅ 

> [The first] commit adds a `style-bare.scss` stylesheet with just the CSS resets and color schemes, without Calypso-opinionated styles.
>
> [The t]hird commit adds the actual entrypoint. [...] It renders the [...] `<Gutenboard />` component. There is no Redux at all in this entrypoint, it just bootstraps the user (so that `user.get()` returns the logged-in user) and then renders a React component. It also doesn't use the Calypso `<Layout>` component. Renders directly into the `wpcom` div.
>
> One thing that doesn't work is the loading placeholder that server sends and browser displays while JS is loading. It's unstyled, because its styles are part of the Calypso-opinionated `main` stylesheet.

One major addition on top of Jarda's code is moving the Gutenboarding code from `client/gutenboarding/` into `client/landing/gutenboarding/`, to bring it in line with other separate entrypoints, and consequently removing the 'hard-wired' section definition from `sections.js` (since those are sections that are by definition for the main entrypoint).

#### Testing instructions

Visit `calypso.localhost:3000/gutenboarding`. Verify that it still works.

Note that some styling looks better (dropdown caret), while some other looks worse (fonts). This is due to the absence of some Calypso 'basic' styling, which was one of the reasons why we went for this approach -- to be able to start from a cleaner slate with regard to styling. We'll work on adding back the bits that we want.

jsnajdr committed:
Add style-bare.scss stylesheet with just the reset and no Calypso styles

jsnajdr committed:
Add Gutenbearding section

ockham committed:
Import style-bare.scss from style.scss

ockham committed:
s/gutenbearding/gutenboarding/g

ockham committed:
s/gutenbeard/gutenboard/g

ockham committed:
Rename folder

ockham committed:
Move Gutenboarding section definition to a constant, remove from sections.js

ockham committed:
Dont makeLayout

ockham committed:
Harmonize

ockham committed:
Remove landing/gutenboarding

ockham committed:
Move client/gutenboarding/ to client/landing/gutenboarding/

ockham committed:
Move gutenboarding section definition to separate file

ockham commented on client/landing/gutenboarding/controller.tsx:
```diff
@@ -19,12 +18,7 @@ export const redirectIfNotEnabled: PageJS.Callback = ( context, next ) => {
-	context.store.dispatch( hideMasterbarAction() );
-	next();
-};
-
```
No longer needed, see https://github.com/Automattic/wp-calypso/pull/37141/files#diff-10f972f707d245f04b2a0136295ff63eR36

ockham commented on client/landing/gutenboarding/controller.tsx:
```diff
@@ -19,12 +18,7 @@ export const redirectIfNotEnabled: PageJS.Callback = ( context, next ) => {
-
 export const main: PageJS.Callback = ( context, next ) => {
-	context.primary = <Gutenboard />;
+	context.layout = <Gutenboard />;
```
`context.layout` gets rendered directly to `#wpcom`. This allows us to omit the actual `<Layout />` component (and masterbar).

simison commented:
Some generic thoughts that we don't necessarily need to solve in this PR:

* How do we bring back cookiebanner? (Assuming it's required)
    <img width="600" alt="image" src="https://user-images.githubusercontent.com/87168/67858125-1cf7b880-fb21-11e9-87b7-bb975c3c41a2.png">

* Do we need to show the help bubble in the bottom corner of the screen? If so, how?
    <img width="72" alt="image" src="https://user-images.githubusercontent.com/87168/67858159-37319680-fb21-11e9-8562-03d548030acc.png">

simison committed:
Move Gutenberg base styles to `/assets/stylesheets`

simison committed:
Remove Calypso overrides

simison committed:
Create more specific CSS entrypoint

- Ensures CSS baseline reset and Gutenboarding base styles reset gets applied in right order since they're in same CSS specificity.
- Add base styles for html & body elements, not included in Gutenberg reset
- Don't include calypso-color-schemes or vendor CSS since we're not using those at the moment.

simison committed:
Put back Calypso color scheme — needed for dev badge! 🤦‍♂️

simison commented on client/assets/stylesheets/style-bare.scss:
```diff
@@ -0,0 +1,8 @@
+@import 'vendor';
+
+// Color Schemes
+@import '~@automattic/calypso-color-schemes/src/calypso-color-schemes';
```
We need these now _only_ for dev badge's colors... should we import it conditionally only in development-mode? That might lead to surprise breakage in production.

Performance penalty including them in the bundle even if we don't use them isn't huge so prolly ok to leave.

More over, there's a good chance that we'll end up using them eventually even if we'd mostly rely on Gutenberg color variables otherwise.

simison commented on client/assets/stylesheets/gutenboarding.scss:
```diff
@@ -0,0 +1,25 @@
+//@import 'vendor';
+
+// Color Schemes
+@import '~@automattic/calypso-color-schemes/src/calypso-color-schemes';
```
We need these now _only_ for dev badge's colors... should we import it conditionally only in development-mode? That might lead to surprise breakage in production.

Performance penalty including them in the bundle even if we don't use them isn't huge so prolly ok to leave.

More over, there's a good chance that we'll end up using them eventually even if we'd mostly rely on Gutenberg color variables otherwise.

ockham replied:
Let's at least add a code comment to that effect -- that it's only used for the dev badge ATM :slightly_smiling_face:

simison committed:
Remove unnesseary html,body padding & margin reset. Already in Calypso CSS reset.

simison commented:
>style-bare.scss stylesheet with just the CSS resets and color schemes, without Calypso-opinionated styles.

I do like the idea, however, I removed this in https://github.com/Automattic/wp-calypso/pull/37141/commits/1d4a35ee3e038defec14d0fa31aaebc54cb57ec2 because for Gutenberg section(s) we need [Gutenberg reset](https://github.com/WordPress/gutenberg/blob/d56979d22dbca9ffa056225bd2b9b8cadd7faadf/packages/base-styles/_mixins.scss#L384-L567) as well the [CSS reset from Calypso](https://github.com/Automattic/wp-calypso/blob/e5a36366a8cee1ece4cdd626195fee45d689a8fe/client/assets/stylesheets/shared/_reset.scss).

We need to ensure they're loaded in a specific order because they're at the same CSS specificity level. Currently in Calypso, entry-point loads somehow after Gutenboarding styles. That's surprising but it seems to be the case in `master` too.

I also wanted to have more control over "bare-styles" as I initially was hoping to include only Calypso reset and not include "vendor" and "calypso-color-schemes" at all (we're not using those yet). I had to add color-schemes back so that dev-badge wouldn't break, so maybe we can still have the bare-styles and just import it in gutenboarding entrypoint.

Lastly, I didn't create a generic "gutenberg" entry-point for now since we only have Gutenboarding using it and I wanted the filename to be clear. If we ever have multiple "Gutenberg entry-points", we can rename.

ockham commented on client/assets/stylesheets/gutenboarding.scss:
```diff
@@ -0,0 +1,23 @@
+// External Dependencies
```
Should this file live here? Or should we colocate it more with the code it is relevant for (i.e. move it to `client/landing/gutenboarding/`), in the spirit of breaking `client/assets/stylesheets/` down into separate units as much as possible (to separate concerns and make clear what is used for what)?

simison replied:
I agree entry-points shouldn't be in `client/assets/stylesheets/*`. 👍

simison replied:
@jsnajdr did you have thoughts on entry-file locations?

ockham commented on client/assets/stylesheets/style.scss:
```diff
@@ -6,6 +6,8 @@
 // Shared
 @import 'shared/reset'; // css reset before the rest of the styles are defined
+
+// Shared
```
I think these two lines can be removed now -- there's a `// Shared` heading on L7 already :slightly_smiling_face:

ockham commented on client/assets/stylesheets/gutenboarding.scss:
```diff
@@ -0,0 +1,23 @@
+// Not an actual "CSS reset", but an opiniated base style for HTML elements in Gutenberg views.
+:root {
+	@include reset;
+}
```
Do we need this reset? How does it compare/relate to [the one in `client/landing/gutenboarding/style.scss`](https://github.com/Automattic/wp-calypso/pull/37141/files#diff-1762b786f4f0e0659667b586a2158493R10)?

simison replied:
Oh yeah, we need to remove the one from Gutenboarding or move both resets there. Looks like I forgot to do that.

The thing is that Calypso reset (line 8 here) and Calypso reset need to be in the same file to ensure they load in that order because they overwrite each other.

simison replied:
Removed duplicate in https://github.com/Automattic/wp-calypso/pull/37141/commits/908fe57685a38efc67d1c4738f31e5fabb171b2d

ockham reviewed (commented):
The `client/gutenboarding/gutenberg-styles/styles.scss → client/assets/stylesheets/gutenberg-base-styles/style.scss` currently breaks the build, since our `Header`'s stylesheet is still referring to the old location: https://github.com/Automattic/wp-calypso/blob/8aa212c0c8c1e64a774d2206ee8878ae73314093/client/gutenboarding/components/header/style.scss#L1

The following patch fixes this

```diff
diff --git a/client/landing/gutenboarding/components/header/style.scss b/client/landing/gutenboarding/components/header/style.scss
index 4ad3d82bba..bf8a0aad3f 100644
--- a/client/landing/gutenboarding/components/header/style.scss
+++ b/client/landing/gutenboarding/components/header/style.scss
@@ -1,4 +1,4 @@
-@import '../../gutenberg-styles/styles';
+@import 'assets/stylesheets/gutenberg-base-styles/style.scss';
 
 // Copied from https://github.com/WordPress/gutenberg/blob/8c0c7b7267473b5c197dd3052e9707f75f4e6448/packages/edit-widgets/src/components/header/style.scss
 .gutenboarding__header {
```

(I hope it's the correct fix :sweat_smile: )

ockham commented:
Thanks a lot @simison , it's looking pretty great -- both in terms of styling, and I think the way the files are organized is also more intuitive now :tada: 

Let's get the relevant changes in and then merge :+1:

ockham commented:
> Some generic thoughts that we don't necessarily need to solve in this PR:
> 
> * How do we bring back cookiebanner? (Assuming it's required)
>   <img alt="image" width="600" src="https://user-images.githubusercontent.com/87168/67858125-1cf7b880-fb21-11e9-87b7-bb975c3c41a2.png">

So far, we aren't even using cookies in Gutenboarding, are we? :grimacing: 

> * Do we need to show the help bubble in the bottom corner of the screen? If so, how?
>   <img alt="image" width="72" src="https://user-images.githubusercontent.com/87168/67858159-37319680-fb21-11e9-8562-03d548030acc.png">

Also not a high-priority thing IMO. There are design/product aspects to this question -- do we offer support (livechat? forums?) during onboarding? I'm not sure if we do during our current NUX flow.

The generic answer to re-using things from Calypso will probably be to isolate them in `@automattic/components`. This will be relatively easy for dumb components (cookie banner?) but very hard for components that need more context/state -- I think the help bubble checks current plan, availability of livechat support staff, etc. So we can't just isolate that in a `components` package; instead, we'd need to tackle hard questions about sharing/encapsulating Calypso state etc.

simison committed:
Remove stray gutenberg CSS reset

simison commented:
> > How do we bring back cookiebanner? (Assuming it's required)
>
>So far, we aren't even using cookies in Gutenboarding, are we? 😬

We'll for sure collect analytics so I think we need it.

simison committed:
Add comment about color scheme

sirreal commented on client/landing/gutenboarding/section.ts:
```diff
@@ -0,0 +1,8 @@
+	paths: [ '/gutenboarding' ],
+	module: 'gutenboarding',
+	secondary: false,
+	group: 'sites',
```
Let's move to its own group:

```suggestion
	group: 'gutenboarding',
```

I'm not sure what all the implications are.

ockham replied:
> I'm not sure what all the implications are.

Probably not too big -- this is mostly about highlighting UI elements in the masterbar IIRC -- which is a non-entity for us.

simison committed:
Comments cleanup

simison commented on client/landing/gutenboarding/section.ts:
```diff
@@ -0,0 +1,8 @@
+	name: 'gutenboarding',
+	paths: [ '/gutenboarding' ],
+	module: 'gutenboarding',
+	secondary: false,
```
`secondary: false` might not be needed here since we don't have Calypso `Layout`?

simison committed:
Add clarifying comment

simison commented on client/landing/gutenboarding/index.ts:
```diff
@@ -0,0 +1,31 @@
+ * Style dependencies
+ */
+import 'assets/stylesheets/gutenboarding.scss';
+import 'components/environment-badge/style.scss';
```
I would expect this style to be applied in every CSS entry point always, and not needing to summon it separately. Good as-is for now tho but just something to maybe refactor in future. cc @jsnajdr

jsnajdr replied:
The `EnvironmentBadge` component is special, and might be a good candidate for completely inlining its styles. Either in the `index.html` document returned from server, or as CSS-in-JS.

simison replied:
Yup, sounds still good for another PR. It also needs little more styling work to look sharp without Calypso base styles:

https://github.com/Automattic/wp-calypso/pull/37141#issuecomment-548256106

simison committed:
Fix Gutenberg base styles import path

simison commented:
Dev badge is depending on some Calypso base styles to look sharp — in Gutenboarding:

<img width="314" alt="image" src="https://user-images.githubusercontent.com/87168/67929230-7dd8cc80-fbc5-11e9-8645-431f811f59da.png">

Elsewhere:

<img width="413" alt="image" src="https://user-images.githubusercontent.com/87168/67929253-8df0ac00-fbc5-11e9-950d-cdb5f726ac87.png">

Ok to address in separate PR.

simison commented:
We'll need to address these:
```
org.eslint.@typescript-eslint/no-explicit-any - /home/circleci/wp-calypso/client/landing/gutenboarding/devtools.ts

line 2, col 34, Warning - Unexpected any. Specify a different type. (@typescript-eslint/no-explicit-any)
```
```
org.eslint.no-restricted-imports - /home/circleci/wp-calypso/server/pages/index.js

'superagent' import is restricted from being used. Please use native `fetch` instead.
```

jsnajdr commented:
> Dev badge is depending on some Calypso base styles to look sharp — in Gutenboarding:

Yes, we had similar issues in the Login Entrypoint few weeks ago.

simison reviewed (approved):
This tests well and I'm quite happy with how it's structured right now. Just small changes and this will be ready.

Small changes might need to be applied to section config: https://github.com/Automattic/wp-calypso/pull/37141#discussion_r340994663 and https://github.com/Automattic/wp-calypso/pull/37141#pullrequestreview-309694110

Linting needs fixing: https://github.com/Automattic/wp-calypso/pull/37141#issuecomment-548258600

I'd leave these for follow-up PRs:
- Possibly move `gutenboarding.scss` and other entry-points out from `assets/stylesheets`
- Address dev badge styling issues (https://github.com/Automattic/wp-calypso/pull/37141#discussion_r340997867, https://github.com/Automattic/wp-calypso/pull/37141#issuecomment-548256106)
- Cookiebanner
- Confirm/fix feature flags (https://github.com/Automattic/wp-calypso/pull/37141#issuecomment-548306103)
- Help badge

simison commented:
Lack of "preferences" and "abtests" sections in dev badge (see https://github.com/Automattic/wp-calypso/pull/37141#issuecomment-548256106) makes me think those might not work in this new entry-point without further work?

jsnajdr commented:
> Lack of "preferences" and "abtests" sections in dev badge makes me think those might not work in this new entry-point without further work?

In the login entrypoint, I had to add the rendering code for the helpers in this commit:
https://github.com/Automattic/wp-calypso/pull/33642/commits/bd6d2fa1771b85b6e45fdaeba36f1784462406e1

The server-generated HTML contains placeholders for them, and the client code needs to render them into the placeholders, as independent React mount roots.

One thing that's worth checking is if the Preferences and ABTest helpers make any sense at all in the Gutenboarding app.

ockham committed:
Move gutenboarding to its own group

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

ockham commented:
> We'll need to address these:
> 
> ```
> org.eslint.@typescript-eslint/no-explicit-any - /home/circleci/wp-calypso/client/landing/gutenboarding/devtools.ts
> 
> line 2, col 34, Warning - Unexpected any. Specify a different type. (@typescript-eslint/no-explicit-any)
> ```

That's a TypeScript error, and it must've existed before this PR, since we're only moving the file around but not otherwise touching it.

> ```
> org.eslint.no-restricted-imports - /home/circleci/wp-calypso/server/pages/index.js
> 
> 'superagent' import is restricted from being used. Please use native `fetch` instead.
> ```

This is pre-existing and fine to leave IMO. Sergio is working on getting rid of these; it's also him who introduced the lint warning to discourage people from using it IIRC.

ockham committed:
Move remaining files

simison commented:
Might be config feature flags at least from URL don't work in the new entry point:

https://wordpress.com/gutenboarding?flags=gutenboarding

@jsnajdr have you noticed the same with login entry point?

jsnajdr commented:
> Might be config feature flags (docs) at least from URL don't work in the new entry point:

I never tested it in the Login Entrypoint, but it should work: the flags are read from `document.location` at the top level of `client/config/index.js`. You don't need to do anything else beyond importing the module.
